### ISSUE TITLE
Exclude files from triggering tsc-watch reload

### ISSUE THREAD

**Issue by chapati23** [gilamran/tsc-watch#89]
I have a little node package that (among other things) reads a JSON schema file and auto-generates a `types.d.ts` file from it.

The problem I have is that `tsc-watch` now thinks that because `types.d.ts` changed, it should reload. Which then again triggers a write to `types.d.ts` and `tsc-watch` is stuck in an infinite restart.

On the other hand, I can't just `exclude` this auto-generated file in my `tsconfig.json` either because I'm using it in the app to strongly type JSON objects.

Is there any way I can have `tsc-watch` ignore a particular file?

**Comment by gilamran** [comment]
I'm not sure that I follow...
You have something that generates a `type.d.ts` and you want to use this file as part of your compilation, that's cool and ok. 
It's good that `tsc-watch` wants to recompile because a file was changed, no matter is an automated process did that or a human...
The thing that I don't understand is why the `types.d.ts` is regenerated after compilation is done...
Also, how is it useful if it's changed AFTER the compilation?

**Comment by chapati23** [comment]
Hey @gilamran, thanks for responding so quickly.
To add a bit more context, I'm working on a side project called https://strong-config.dev/

Overly simplified: it loads a config file based on your current env (`dev`, `staging`, `prod` etc).
To make config files more robust, we allow strongly typing them by providing a JSON Schema that defines your config shape. We use JSON Schema because
a) we don't want to force a consumer of this package to use TypeScript.
b) we want to be able to validate config files at runtime (where typescript isn't available)

However, for TypeScript users, we have a feature that allows consuming your `schema.json` and auto-generating a `config.d.ts` (what I called `types.d.ts` above for clarity). For convenience (in development mode only), we re-generate this file every time the app starts so the schema file is always in sync with the types file.

The main purpose for the auto-generated `config.d.ts` file is to get better IntelliSense for the `config` object in your editor.

> The thing that I don't understand is why the types.d.ts is regenerated after compilation is done...
tl;dr: Because it's not TypeScript that's generating the file but the application itself

**In non-TS projects we are using `nodemon`, which has its own `ignore: []` option, which is what would be wonderful to have in `tsc-watch`**

**Comment by gilamran** [comment]
> we re-generate this file every time the app starts so the schema file is always in sync with the types file.

This sounds like this schema generation should happen BEFORE the compilation starts, what's the use of a `config.d.ts` if it's not really used in the compilation. And probably the first compilation will fail, as the generated type does not exist.

I think that you should have a separated process that watches the schema, and generates the `config.d.ts`, if that file changes a new compilation should start.

Doing the same with `ts-node` or typescript itself will cause the same issue.

Anyways, `tsc-watch` is not the one that's listening to file changes, it's typescript itself (I'm forcing the `--watch` parameter), the only thing that `tsc-watch` does is running the onSuccess or onFail actions.

I'll be glad to help, but I don't see what I can do to help in situation.
Let me know.

**Comment by chapati23** [comment]
I agree that frontloading the schema generation would be cleanest to avoid this recompile loop of death. However, the DX gets a lot worse for our package then because users would have to manually wire up and adapt the schema generation to their respective development process. Needing two separate watcher processes is probably too cumbersome for most people just to use a library.

That's why we made the decision to have it happen "automagically" 🧙‍♂️

The one other possible solution could be to write the auto-generated types file into `node_modules/@strong-config/node` somehow which should circumvent the endless-recompile but will probably open up another can of worms with TypeScript applying the right types..

However, to stop stealing your time here I think you're right in that this is more of a `tsc` issue than `tsc-watch` in particular. I drew the wrong conclusion that `nodemon === tsc-watch` but they're actually more different than I thought initially.

FWIW, We don't seem to be the only ones having this problem: https://github.com/microsoft/TypeScript/issues/27279

**Comment by gilamran** [comment]
I hope that you'll have a good solution soon.

**Comment by divmgl** [comment]
So what was the resolution on this? I'm experiencing this now. How do we get `tsc-watch` to only watch `.ts` files? Or potentially exclude all other files?

**Comment by gilamran** [comment]
@divmgl what is your specific issue? 
Note that `tsc-watch` is using `tsc` to do the compilation, but only add a `--watch` flag to force watch. So, if it's recompiling it's because `tsc` detected a change... maybe you should add the files that you want to ignore to the `exclude` on `tsconfig.json` file.

**Comment by divmgl** [comment]
@gilamran this is starting to happen very frequently. It isn't happening to us using the regular `tsc --watch` so the issue is in this package. We're alt-tabbing back into VSCode and `tsc` is saying a change has been detected. I have to emphasize that this is _not_ happening when running `tsc --watch` normally.

Additionally, `tsc-watch` is not properly stopping the commands. Even on the newest `dev` branch this issue is happening. I'm having to `lsof -i :3000` to kill the server.

**Comment by divmgl** [comment]
I'm going to open a different issue as I think it's different from this issue.
